File: python/leetcode/267.py
class Solution(object):
    def generatePalindromes(self, s):
        res = []
        t, mid = [], []
        words = collections.defaultdict(int)
        for ch in s: words[ch] += 1
        for key, val in words.iteritems():
            if val % 2 == 1: mid.append(key)
            if len(mid) > 1: return []
            t.extend(list(key * (val/2)))
        self.permute(t, 0, mid, res)
        return res

    def permute(self, t, start, mid, res):
        if start >= len(t):
            res.append("".join(t + mid + t[::-1]))
            return 
        for i in range(start, len(t)):
            if i != start and t[i] == t[start]: continue
            t[i], t[start] = t[start], t[i]
            self.permute(t, start + 1, mid, res)
            t[i], t[start] = t[start], t[i]

File: cpp/leetcode/267.cpp
class Solution {
public:
    vector<string> generatePalindromes(string s) {
        unordered_map<char, int> words;
        for(auto ch : s) words[ch]++;
        string part = "", mid = "";
        for(auto it : words){
            if(it.second % 2 == 0) {
                part += string(it.second/2, it.first);
            } else {
                mid += it.first;
                if(mid.size() > 1) return {};
                part += string(it.second/2, it.first);
            }
        }
        vector<string> res;
        dfs(0, mid, part, res);
        return res;
    }
    
    void dfs(int start, string mid, string &part, vector<string> &res){
        if(start == part.size()){
            res.push_back(part + mid + string(part.rbegin(), part.rend()));
            return;
        }
        for(int i = start; i < part.size(); i++){
            if(i > start && part[i] == part[start]) continue;
            swap(part[start], part[i]);
            dfs(start + 1, mid, part, res);
            swap(part[start], part[i]);
        }
    }
};